packages/opentelemetry-plugin-mongodb/src/mongodb.ts

```diff
@@ -0,0 +1,255 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// mongodb.Server type is deprecated so every use trigger a lint error
+/* tslint:disable:deprecation */
+
+import { BasePlugin, NoopLogger } from '@opentelemetry/core';
+import {
+  Span,
+  SpanKind,
+  Logger,
+  Tracer,
+  CanonicalCode,
+} from '@opentelemetry/types';
+import * as mongodb from 'mongodb';
+import * as shimmer from 'shimmer';
+
+type Func<T> = (...args: unknown[]) => T;
+type MongoDB = typeof mongodb;
+interface MongoInternalCommand {
+  findandmodify: boolean;
+  createIndexes: boolean;
+  count: boolean;
+  ismaster: boolean;
+}
+
+/** MongoDB instrumentation plugin for OpenTelemetry */
+export class MongoDBPlugin extends BasePlugin<MongoDB> {
+  private readonly SERVER_FNS = ['insert', 'update', 'remove'];
+  private readonly CURSOR_FNS_FIRST = ['_find', '_getmore'];
+  protected _logger!: Logger;
+  protected readonly _tracer!: Tracer;
+
+  constructor(public moduleName: string) {
+    super();
+    // TODO: remove this once a logger will be passed
+    // https://github.com/open-telemetry/opentelemetry-js/issues/193
+    this._logger = new NoopLogger();
+  }
+
+  /**
+   * Patches MongoDB operations.
+   */
+  protected patch() {
+    this._logger.debug('Patched MongoDB');
+
+    if (this._moduleExports.Server) {
+      this._logger.debug('patching mongodb-core.Server.prototype.command');
+      shimmer.wrap(
+        this._moduleExports.Server.prototype,
+        'command' as never,
+        // tslint:disable-next-line:no-any
+        this.getPatchCommand() as any
+      );
+      this._logger.debug(
+        'patching mongodb-core.Server.prototype functions:',
+        this.SERVER_FNS
+      );
+      shimmer.massWrap(
+        [this._moduleExports.Server.prototype],
+        this.SERVER_FNS as never[],
+        // tslint:disable-next-line:no-any
+        this.getPatchQuery() as any
+      );
+    }
+
+    if (this._moduleExports.Cursor) {
+      this._logger.debug(
+        'patching mongodb-core.Cursor.prototype functions:',
+        this.CURSOR_FNS_FIRST
+      );
+      shimmer.massWrap(
+        [this._moduleExports.Cursor.prototype],
+        this.CURSOR_FNS_FIRST as never[],
+        // tslint:disable-next-line:no-any
+        this.getPatchCursor() as any
+      );
+    }
+
+    return this._moduleExports;
+  }
+
+  /** Unpatches all MongoDB patched functions. */
+  unpatch(): void {
+    shimmer.unwrap(this._moduleExports.Server.prototype, 'command' as never);
+    shimmer.massUnwrap([this._moduleExports.Server.prototype], this
+      .SERVER_FNS as never[]);
+    shimmer.massUnwrap([this._moduleExports.Cursor.prototype], this
+      .CURSOR_FNS_FIRST as never[]);
+  }
+
+  /** Creates spans for Command operations */
+  private getPatchCommand() {
+    const plugin = this;
+    return (original: Func<mongodb.Server>) => {
+      return function(
```

Comment:
Do you think that we should give a name to this function ? This would help when profiling. WDYT?